### PR TITLE
add init_score & test cpp and python result consistency

### PR DESCRIPTION
**Pull request by wxchan**
@guolinke my new added test cases always fail. Can you help check where goes wrong?

### CODE REVIEW

**Comment by guolinke** [comment]
@wxchan sure, I will check it

**Comment by guolinke** [comment]
@wxchan I think the reason is the different parameters.
I think you can simply use a toy example for the test, like this: https://github.com/Microsoft/LightGBM/blob/master/tests/python_package_test/test_engine.py#L213-L245

**Comment by wxchan** [comment]
@guolinke but I read params from train.conf files, it should be the same. The strange thing is only a few of samples fail.

**Comment by guolinke** [comment]
@wxchan i found the trained the model is exact the same, but the prediction is not. 
Maybe there is something wrong in prediction

**Comment by wxchan** [comment]
@guolinke maybe something to do with those fields like init_score? It passed when I removed init_score.

**Comment by guolinke** [comment]
@wxchan really ?  I think the init_score should have no impact on prediction.

**Comment by guolinke** [comment]
@wxchan 
I found the problem is caused by the Atof: https://github.com/Microsoft/LightGBM/blob/master/include/LightGBM/utils/common.h#L163-L251

if replaced it with the std::stod, everything seems fine. 
```
inline static const char* Atof(const char* p, double* out) {
  *out = NAN;
  while (*p == ' ') {
    ++p;
  }
  size_t cnt = 0;
  while (*(p + cnt) != '\0' && *(p + cnt) != ' '
         && *(p + cnt) != '\t' && *(p + cnt) != ','
         && *(p + cnt) != '\n' && *(p + cnt) != '\r'
         && *(p + cnt) != ':') {
    ++cnt;
  }
  if (cnt > 0) {
    std::string str(p, cnt);
    p += cnt;
    std::transform(str.begin(), str.end(), str.begin(), Common::tolower);
    if (str == std::string("inf")) {
      *out = 1e308;
    } else if (str == std::string("-inf")) {
      *out = -1e308;
    } else if (str != std::string("na") && str != std::string("nan")) {
      *out = std::stod(str);
    }
  }
  while (*p == ' ') {
    ++p;
  }
  return p;
}
```

**Comment by guolinke** [comment]
@wxchan however, stod seems very slow: https://tinodidriksen.com/2011/05/cpp-convert-string-to-double-speed/

**Comment by wxchan** [comment]
@guolinke does it cause by some feature values near thresholds fall into different sub-trees because the precision of model?

**Comment by guolinke** [comment]
@wxchan yeah, exactly. I am trying to update atof .

**Comment by guolinke** [comment]
@wxchan can you try to replace the Atof with:

```
template<class T>
inline static T Pow(T base, int power) {
  if (power == 0) {
    return 1;
  } else if (power == 1) {
    return base;
  } else if (power % 2 == 0) {
    return Pow(base*base, power / 2);
  } else if (power % 3 == 0) {
    return Pow(base*base*base, power / 3);
  } else {
    return base * Pow(base, power - 1);
  }
}

inline static const char* Atof(const char* p, double* out) {
  int frac;
  double sign, value, scale;
  *out = NAN;
  // Skip leading white space, if any.
  while (*p == ' ') {
    ++p;
  }

  // Get sign, if any.
  sign = 1.0;
  if (*p == '-') {
    sign = -1.0;
    ++p;
  } else if (*p == '+') {
    ++p;
  }

  // is a number
  if ((*p >= '0' && *p <= '9') || *p == '.' || *p == 'e' || *p == 'E') {
    // Get digits before decimal point or exponent, if any.
    for (value = 0.0; *p >= '0' && *p <= '9'; ++p) {
      value = value * 10.0 + (*p - '0');
    }

    // Get digits after decimal point, if any.
    if (*p == '.') {
      double right = 0.0;
      int nn = 0;
      ++p;
      while (*p >= '0' && *p <= '9') {
        right = (*p - '0') + right * 10.0;
        ++nn;
        ++p;
      }
      value += right / Pow(10.0, nn);
    }

    // Handle exponent, if any.
    frac = 0;
    scale = 1.0;
    if ((*p == 'e') || (*p == 'E')) {
      uint32_t expon;
      // Get sign of exponent, if any.
      ++p;
      if (*p == '-') {
        frac = 1;
        ++p;
      } else if (*p == '+') {
        ++p;
      }
      // Get digits of exponent, if any.
      for (expon = 0; *p >= '0' && *p <= '9'; ++p) {
        expon = expon * 10 + (*p - '0');
      }
      if (expon > 308) expon = 308;
      // Calculate scaling factor.
      while (expon >= 50) { scale *= 1E50; expon -= 50; }
      while (expon >= 8) { scale *= 1E8;  expon -= 8; }
      while (expon > 0) { scale *= 10.0; expon -= 1; }
    }
    // Return signed and scaled floating point result.
    *out = sign * (frac ? (value / scale) : (value * scale));
  } else {
    size_t cnt = 0;
    while (*(p + cnt) != '\0' && *(p + cnt) != ' '
      && *(p + cnt) != '\t' && *(p + cnt) != ','
      && *(p + cnt) != '\n' && *(p + cnt) != '\r'
      && *(p + cnt) != ':') {
      ++cnt;
    }
    if (cnt > 0) {
      std::string tmp_str(p, cnt);
      std::transform(tmp_str.begin(), tmp_str.end(), tmp_str.begin(), Common::tolower);
      if (tmp_str == std::string("na") || tmp_str == std::string("nan")) {
        *out = NAN;
      } else if (tmp_str == std::string("inf") || tmp_str == std::string("infinity")) {
        *out = sign * 1e308;
      } else {
        Log::Fatal("Unknown token %s in data file", tmp_str.c_str());
      }
      p += cnt;
    }
  }

  while (*p == ' ') {
    ++p;
  }

  return p;
}
```

**Comment by guolinke** [comment]
strange, I can pass the test of binary classification test locally

**Comment by guolinke** [comment]
@wxchan did chech the dtypes of np.loadtext? it is float64 in my local machine.

**Comment by wxchan** [comment]
@guolinke it's float64 on my machine too, but it still fails. It's possible something to do with np.loadtext because loading with load_svmlight_file passes the test.

**Comment by wxchan** [comment]
@guolinke still fail if I init dataset with file path string.

**Comment by guolinke** [comment]
@wxchan I can pass these tests except multi-class in my local machines.

**Comment by guolinke** [comment]
@wxchan refer to the test result: https://travis-ci.org/Microsoft/LightGBM/jobs/292447965 
I think the different is very minor. So maybe we can use other solution to fix the test.
BTW, i test the speed of strtod, it is about 1.5x slow ...  So I think we should not switch to it.

**Comment by guolinke** [comment]
@wxchan I don't why these tests are still failed: https://travis-ci.org/Microsoft/LightGBM/jobs/292509708

they all pass in my local machine

**Comment by wxchan** [comment]
@guolinke how about giving some tolerances of inequality? I think both results are reasonable?

**Comment by guolinke** [comment]
@wxchan I think we can check the model, instead of the prediction result

**Comment by wxchan** [comment]
@guolinke then we can't check consistency of prediction code.

**Comment by guolinke** [comment]
@wxchan we can check it by using the same model. I remember python-package can predict from file, which calls the same function of cpp.

**Comment by wxchan** [comment]
@guolinke I tried that, strangely can't pass tests too.

**Comment by guolinke** [comment]
still the float point issue? did you try the std::atof?

**Comment by guolinke** [comment]
@wxchan I think my PR on your branch should pass the tests. 
you can try it with the model comparison.

File: .travis/test.sh [hunk 1]
```diff
@@ -115,7 +115,12 @@ else
     cmake ..
 fi
 
-make _lightgbm || exit -1
+make || exit -1
+
+cd $TRAVIS_BUILD_DIR/examples/binary_classification && ../../lightgbm config=train.conf && ../../lightgbm config=predict.conf || exit -1
+cd $TRAVIS_BUILD_DIR/examples/regression && ../../lightgbm config=train.conf && ../../lightgbm config=predict.conf || exit -1
+cd $TRAVIS_BUILD_DIR/examples/multiclass_classification && ../../lightgbm config=train.conf && ../../lightgbm config=predict.conf || exit -1
+cd $TRAVIS_BUILD_DIR/examples/lambdarank && ../../lightgbm config=train.conf && ../../lightgbm config=predict.conf || exit -1
```
Comment (guolinke) [inline .travis/test.sh]: these can be remove ?

File: python-package/lightgbm/basic.py [hunk 2]
```diff
@@ -568,14 +568,16 @@ def __init__(self, data, label=None, max_bin=255, reference=None,
             If string, it represents the path to txt file.
         label : list or numpy 1-D array, optional (default=None)
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py]: Please add possible type None
`label : list, numpy 1-D array or None, optional (default=None)`

File: python-package/lightgbm/basic.py [hunk 3]
```diff
@@ -568,14 +568,16 @@ def __init__(self, data, label=None, max_bin=255, reference=None,
...
             Label of the data.
-        max_bin : int, optional (default=255)
+        max_bin : int, optional (default=None)
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py]: Please add possible type None
max_bin : int or None, optional (default=None)

File: python-package/lightgbm/basic.py [hunk 4]
```diff
@@ -568,14 +568,16 @@ def __init__(self, data, label=None, max_bin=255, reference=None,
...
             Max number of discrete bins for features.
         reference : Dataset or None, optional (default=None)
             If this is Dataset for validation, training data should be used as reference.
         weight : list, numpy 1-D array or None, optional (default=None)
             Weight for each instance.
         group : list, numpy 1-D array or None, optional (default=None)
             Group/query size for Dataset.
+        init_score : list, numpy 1-D array or None, optional (default=None)
+            Init_score for Dataset.
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py]: Please remove underscore.

File: python-package/lightgbm/basic.py [hunk 5]
```diff
@@ -842,6 +854,8 @@ def create_valid(self, data, label=None, weight=None, group=None,
...
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py]: Please remove underscore.

**Comment by guolinke** [comment]
@wxchan updates ?

**Comment by wxchan** [comment]
@guolinke I will add consistency check of sklearn interface later, maybe weekend.

**Comment by chivee** [comment]
related to #1033

**Comment by wxchan** [comment]
@guolinke @StrikerRUS can we set sklearn default random_state=None, and filter it out if it's None? there exists some inconsistency now.

**Comment by StrikerRUS** [comment]
@wxchan I think that setting any parameter to `None` without the appropriate description (except the case when `None` means no using this parameter) very confusing.

To speak about `random_state`, can we place sklearn's specific parameters into cpp `alias_table` too?

**Comment by wxchan** [comment]
@StrikerRUS it makes no difference. The difference now is, the seed in native api has no default value, it will use c++ default *_seed; sklearn api has a default seed=0, so those *_seed will be set to some other values.

I think users just want reproducible behavior by using seed, they just need to know it's fixed result or random.

**Comment by wxchan** [comment]
speak of this, a totally different subject: there was a feature request about supporting numpy RandomState for random_state, I think it's unnecessary and didn't have a good idea for it, perhaps we can support it if it's simple. https://github.com/Microsoft/LightGBM/issues/730

**Comment by StrikerRUS** [comment]
@wxchan My first paragraph of comment was more about `max_bin` and other params.

Your idea about ignoring `random_state` if it's `None` and usind default `*_seed` values looks good to me
>set sklearn default random_state=None, and filter it out if it's None

**Comment by wxchan** [comment]
@StrikerRUS how about just delete it from `__init__`? It can still be accessed by kwargs.

**Comment by StrikerRUS** [comment]
@wxchan I think it'll be confusing: `random_state` is one of the *standard* (I don't know how to name it...) sklearn's parameters and users will think it's not presented in LightGBM at all if they will not see it in the constructor's signature.

File: python-package/lightgbm/basic.py [hunk 6]
```diff
@@ -555,8 +555,8 @@ def __pred_for_csc(self, csc, num_iteration, predict_type):
 
 class Dataset(object):
     """Dataset in LightGBM."""
-    def __init__(self, data, label=None, max_bin=255, reference=None,
-                 weight=None, group=None, silent=False,
+    def __init__(self, data, label=None, max_bin=None, reference=None,
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py]: May I ask why you've changed `max_bin` to `None` in public interface when it's actually still `255`? I wrote in my previous comment about such cases that they're confusing. If you worry about you'll miss some places while changing the default value in Parameters.rst, just search GitHub with `param=old_value`.

File: python-package/lightgbm/VERSION.txt [hunk 7]
```diff
@@ -0,0 +1 @@
+2.0.10
```
Comment (StrikerRUS) [inline python-package/lightgbm/VERSION.txt]: What for do we need the duplicate of this?
https://github.com/Microsoft/LightGBM/blob/master/VERSION.txt

Comment (wxchan) [reply python-package/lightgbm/VERSION.txt]: I add it to git accidentally

File: python-package/lightgbm/basic.py (hunk 6 above)
Comment (wxchan) [reply python-package/lightgbm/basic.py]: it's set by @guolinke

**Comment by wxchan** [comment]
Done.

**Comment by wxchan** [comment]
the gpu test fails. is there some parameter inconsistency?

**Comment by StrikerRUS** [comment]
@wxchan Maybe because of this?

> * Question 5: When using LightGBM GPU, I cannot reproduce results over several runs.
> * Solution 5: It is a normal issue, there is nothing we/you can do about, you may try to use gpu_use_dp = true for reproducibility (see Microsoft/LightGBM#560). You may also use CPU version.

File: python-package/lightgbm/basic.py (hunk 6 above)
Comment (guolinke) [reply python-package/lightgbm/basic.py]: @StrikerRUS This is to solve problem of the parameter priority. User can set ```max_bin``` (and other parameters) in both the parameter dict and the function arguments. This will cause the problem: we don't know which one is high "priority" when user set it in both two places. As a result, we can set the default function argument to None. If user didn't set the function argument, we use the value in parameter dict, otherwise use the function arguments. 

We can add something like ```defualt=None (255 in c_api) ``` in python doc, to avoid the addiction search of that parameter.

**Comment by guolinke** [comment]
@wxchan maybe you can try run gpu by python two times, and check the consistency. And they are not the same, we can remove the consistency in GPU test.

**Comment by wxchan** [comment]
Fixed by gpu_use_dp=true. @guolinke @StrikerRUS

File: python-package/lightgbm/basic.py (hunk 6 above)
Comment (StrikerRUS) [reply python-package/lightgbm/basic.py]: @guolinke 
I suppose it's obvious for users that function argument shouldn't be duplicated in param dict and parameter in signature which has the default value and the own description has higher priority than parameter in dict.

Maybe than it's better to completely remove `max_been` from arguments (for 1 or 2 release mark as deprecated)? Because it's oddly that `None` actually means `255`...

Comment (guolinke) [reply python-package/lightgbm/basic.py]: some users like to put all parameters into dicts( like me), some like to put them into function arguments. I think set default to None is reasonable, since it is consistent with CLI version.

Comment (guolinke) [reply python-package/lightgbm/basic.py]: I think we can remove them from arguments as well

File: python-package/lightgbm/basic.py [hunk 8]
```diff
@@ -633,7 +636,8 @@ def _lazy_init(self, data, label=None, max_bin=255, reference=None,
         params = {} if params is None else params
         self.max_bin = max_bin
         self.predictor = predictor
-        params["max_bin"] = max_bin
+        if self.max_bin is not None:
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py]: @wxchan Please add deprecation warning here.

>StrikerRUS 23 hours ago Collaborator 
>Maybe than it's better to completely remove max_been from arguments (for 1 or 2 release mark as deprecated)? Because it's oddly that None actually means 255...

>guolinke an hour ago Member 
>I think we can remove them from arguments as well

**Comment by wxchan** [comment]
I am outside. If there are no other issues, can we merge this first and fix it in following commit? @guolinke @StrikerRUS

**Comment by guolinke** [comment]
@wxchan sure.
@StrikerRUS you can open a new issue to track the None and default value problem.

**Comment by StrikerRUS** [comment]
@wxchan @guolinke OK, will do it.

**Comment by wxchan** [comment]
@guolinke merge this now?
